docs/metrics-overview.md

```diff
@@ -0,0 +1,116 @@
+# COSI Driver Metrics Documentation
+
+This document provides an overview of the Prometheus metrics exposed by the COSI driver. These metrics are designed to help open-source users monitor the performance and operations of the COSI driver. The metrics cover gRPC server calls, S3 API operations, and IAM API interactions.
+
+## Metrics Overview
+
+Metrics are exposed at the `/metrics` endpoint on the address configured via the `--metrics-address` flag (default: `:8080`). These metrics are Prometheus-compatible and can be used to create dashboards for observability.
+
+---
+
+## gRPC Metrics
+
+The COSI driver exposes gRPC server metrics to monitor RPC activity.
+
+| Metric Name                     | Description                                                | Labels                                     |
+|---------------------------------|------------------------------------------------------------|--------------------------------------------|
+| `grpc_server_started_total`     | Total number of RPCs started on the server.                | `grpc_method`, `grpc_service`, `grpc_type` |
+| `grpc_server_handled_total`     | Total number of RPCs completed on the server.              | `grpc_method`, `grpc_service`, `grpc_code` |
+| `grpc_server_msg_received_total`| Total number of messages received by the server.           | `grpc_method`, `grpc_service`             |
+| `grpc_server_msg_sent_total`    | Total number of messages sent by the server.               | `grpc_method`, `grpc_service`             |
+| `grpc_server_handling_seconds`  | Time taken for RPC calls to be handled by the server.       | `grpc_method`, `grpc_service`             |
+
+### Example gRPC Methods
+
+- Methods: `DriverCreateBucket`, `DriverDeleteBucket`, `DriverGetInfo`, `DriverGrantBucketAccess`, `DriverRevokeBucketAccess`
+- Services: `cosi.v1alpha1.Provisioner`, `cosi.v1alpha1.Identity`
+
+---
+
+## S3 Metrics
+
+These metrics track the total number of S3 requests and the duration of each request.
+
+| Metric Name                      | Description                                                 | Labels            |
+|----------------------------------|-------------------------------------------------------------|-------------------|
+| `s3_requests_total`              | Total number of S3 requests, categorized by method and status. | `method`, `status` |
+| `s3_request_duration_seconds`    | Duration of S3 requests in seconds, categorized by method and status. | `method`, `status` |
```

Comment:
Type (`Counter`, `Gauge`, `Histogram`) is missing.

Also as they have the same labels, there will be duplication between:
- `s3_requests_total`
- `s3_request_duration_seconds_count`

In S3C to reduce cardinality when it can be high (for like s3 and vault) I started replacing usage of the total counter with the duration histogram `_count` when they have the same labels.